pkg/types/interfacetests/codec_interface_tests.go

```diff
@@ -0,0 +1,377 @@
+			assert.Equal(t, &secondItem, actual)
+		},
+		"GetMaxEncodingSize returns errors for unknown types": func(t *testing.T) {
+			cr := tester.GetChainReader(t)
+			_, err := cr.GetMaxEncodingSize(ctx, 10, "not"+TestItemType)
+			assert.IsType(t, types.InvalidTypeError{}, err)
+		},
+	}
+
+	runTests(t, tester, tests)
+}
+
+// RunChainReaderWithStrictArgsInterfaceTest is meant to be used by codecs that don't pad
+// They can assure that the right argument size is verified.
+// Padding makes that harder/impossible to verify for come codecs.
+// However, the extra verification is nice to have when possible.
+func RunChainReaderWithStrictArgsInterfaceTest(t *testing.T, tester ChainReaderInterfaceTester) {
+	ctx := context.Background()
+	RunChainReaderInterfaceTests(t, tester)
+
+	tests := map[string]func(t *testing.T){
+		"Gives an error decoding extra fields on an item": func(t *testing.T) {
+			item := CreateTestStruct(0, tester.GetAccountBytes)
+			req := &EncodeRequest{
+				TestStructs: []TestStruct{item},
+				ExtraField:  true,
+				TestOn:      TestItemType,
+			}
+			resp := tester.EncodeFields(t, req)
+			codec := tester.GetChainReader(t)
+			err := codec.Decode(ctx, resp, &item, TestItemType)
+			assert.IsType(t, types.InvalidEncodingError{}, err)
+		},
+		"Gives an error decoding missing fields on an item": func(t *testing.T) {
+			item := CreateTestStruct(0, tester.GetAccountBytes)
+			req := &EncodeRequest{
+				TestStructs:  []TestStruct{item},
+				MissingField: true,
+				TestOn:       TestItemType,
+			}
+			resp := tester.EncodeFields(t, req)
+			codec := tester.GetChainReader(t)
+			err := codec.Decode(ctx, resp, &item, TestItemType)
+			assert.IsType(t, types.InvalidEncodingError{}, err)
+		},
+		"Gives an error decoding extra fields on a slice": func(t *testing.T) {
+			items := []TestStruct{CreateTestStruct(0, tester.GetAccountBytes)}
+			req := &EncodeRequest{
+				TestStructs: items,
+				ExtraField:  true,
+				TestOn:      TestItemSliceType,
+			}
+			resp := tester.EncodeFields(t, req)
+			codec := tester.GetChainReader(t)
+			err := codec.Decode(ctx, resp, &items, TestItemSliceType)
+			assert.IsType(t, types.InvalidEncodingError{}, err)
+		},
+		"Gives an error decoding missing fields on an slice": func(t *testing.T) {
+			items := []TestStruct{CreateTestStruct(0, tester.GetAccountBytes)}
+			req := &EncodeRequest{
+				TestStructs:  items,
+				MissingField: true,
+				TestOn:       TestItemSliceType,
+			}
+			resp := tester.EncodeFields(t, req)
+			codec := tester.GetChainReader(t)
+			err := codec.Decode(ctx, resp, &items, TestItemSliceType)
+			assert.IsType(t, types.InvalidEncodingError{}, err)
+		},
+	}
+
+	runTests(t, tester, tests)
+}
+
+func runTests(t *testing.T, tester ChainReaderInterfaceTester, tests map[string]func(t *testing.T)) {
+	for name, test := range tests {
+		t.Run(name, func(t *testing.T) {
+			tester.Setup(t)
+			defer func() { tester.Teardown(t) }()
```

Comment:
Does this need to be a seperate step? Or could `Setup` just register via `t.Cleanup` internally, if necessary?